File: lee/src/main/java/org/example/lee/题目/二叉树/b_二叉搜索树中第k小的元素.java
package org.example.lee.题目.二叉树;

import org.example.lee.model.TreeNode;

public class b_二叉搜索树中第k小的元素 {

    public int num = 0;
    public int kthSmallest(TreeNode root, int k) {
        TreeNode node = tool(root,k);
        return node.val;
    }
    TreeNode tool(TreeNode root,int k){
        if(root == null){
            return null;
        }
        TreeNode a =  tool(root.left,k);
        if(a!=null){
            return a;
        }
        num++;
        if(num == k){
            return root;
        }
        a =tool(root.right,k);
        return a;
    }
}

File: lee/src/main/java/org/example/lee/题目/二叉树/b_二叉搜索树中第K小的元素.go
package 二叉树

func kthSmallest(root *TreeNode, k int) int {
	node, _ := ktool(root, k)
	if node != nil {
		return node.Val
	}
	return -1
}

func ktool(root *TreeNode, k int) (*TreeNode, int) {
	if root == nil {
		return nil, k
	}
	a, b := ktool(root.Left, k)
	if a != nil {
		return a, b
	}
	b--
	if b == 0 {
		return root, b
	}
	return ktool(root.Right, b)
}
